### ISSUE TITLE
"wrong number of arguments for 'mset' command" when using redis

### ISSUE THREAD

**Issue by nchampsavoir** [BertrandBordage/django-cachalot#28]
Hi,

When using redis and Dango 1.8, I've got the following exception : "wrong number of arguments for 'mset' command"

The stack trace shows that in cachalot/transaction.py line 34:
> _invalidate_table_cache_keys(self.parent_cache, self.to_be_invalidated)

self.to_be_invalidated is []

which makes line 133 of cachalot/utils.py 
> cache.set_many(d, None)

raise an exception because d is none and Redis does not support "mset" with no arguments.

Any ideas before I try to investigate further ?

**Comment by BertrandBordage** [comment]
Hello @nchampsavoir,

Never faced that issue, can someone else confirm? I tested with Django 1.8.3, django-redis 4.2.0, redis-py 2.10.3, Redis 2.8.4 and django-cachalot 1.0.3.
Are you using django-redis?

You say that `d` is `None`, leading to this error, but it can never happen. As you can see [in the lines above](https://github.com/BertrandBordage/django-cachalot/blob/4ae8bccf962fd811e851391a091b25486af7814d/cachalot/utils.py#L130-L132), `d` is always a dictionary.

**Comment by eressler** [comment]
Can confirm I am running into this issue currently while attempting to edit an existing item in the django admin.

Running on Heroku using 'Heroku Redis' as the redis provider, following instructions provided here:
https://devcenter.heroku.com/articles/heroku-redis#connecting-in-django
django-redis-cache==1.6.0
Django==1.8.4
django-cachalot==1.0.3

Heroku Redis server info:
```
ec2-54-83-32-60.compute-1.amazonaws.com:9279> info
# Server
redis_version:3.0.2
redis_git_sha1:##########
redis_git_dirty:#
redis_build_id:##########
redis_mode:standalone
os:Linux 3.2.0-85-virtual x86_64
arch_bits:64
multiplexing_api:epoll
gcc_version:4.6.3
process_id:7
run_id:##################
tcp_port:#####
uptime_in_seconds:133928
uptime_in_days:1
hz:10
lru_clock:667041
```

Full stack trace:
```
ResponseError at /admin/contractors/contractor/add/
wrong number of arguments for 'mset' command
Request Method:	GET
Request URL:	http://*******.herokuapp.com/admin/contractors/contractor/add/
Django Version:	1.8.4
Exception Type:	ResponseError
Exception Value:	
wrong number of arguments for 'mset' command
Exception Location:	/app/.heroku/python/lib/python2.7/site-packages/redis/connection.py in read_response, line 574
Python Executable:	/app/.heroku/python/bin/python
Python Version:	2.7.10
Python Path:	
['/app',
 '/app/.heroku/python/bin',
 '/app/.heroku/src/configurations',
 '/app/.heroku/python/lib/python2.7/site-packages/setuptools-18.1-py2.7.egg',
 '/app/.heroku/python/lib/python2.7/site-packages/pip-7.1.0-py2.7.egg',
 '/app',
 '/app/.heroku/python/lib/python27.zip',
 '/app/.heroku/python/lib/python2.7',
 '/app/.heroku/python/lib/python2.7/plat-linux2',
 '/app/.heroku/python/lib/python2.7/lib-tk',
 '/app/.heroku/python/lib/python2.7/lib-old',
 '/app/.heroku/python/lib/python2.7/lib-dynload',
 '/app/.heroku/python/lib/python2.7/site-packages']
Server time:	Tue, 29 Sep 2015 05:55:38 +0000
Traceback Switch to copy-and-paste view

/app/.heroku/python/lib/python2.7/site-packages/django/core/handlers/base.py in get_response
                                response = wrapped_callback(request, *callback_args, **callback_kwargs) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/django/contrib/admin/options.py in wrapper
                            return self.admin_site.admin_view(view)(*args, **kwargs) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/django/utils/decorators.py in _wrapped_view
                                response = view_func(request, *args, **kwargs) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/django/views/decorators/cache.py in _wrapped_view_func
                    response = view_func(request, *args, **kwargs) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/django/contrib/admin/sites.py in inner
                        return view(request, *args, **kwargs) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/django/contrib/admin/options.py in add_view
                    return self.changeform_view(request, None, form_url, extra_context) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/django/utils/decorators.py in _wrapper
                        return bound_func(*args, **kwargs) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/django/utils/decorators.py in _wrapped_view
                                response = view_func(request, *args, **kwargs) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/django/utils/decorators.py in bound_func
                            return func.__get__(self, type(self))(*args2, **kwargs2) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/django/utils/decorators.py in inner
                                return func(*args, **kwargs) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/cachalot/monkey_patch.py in inner
                                                    and not needs_rollback) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/cachalot/cache.py in exit_atomic
                            atomic_cache.commit() ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/cachalot/transaction.py in commit
                    _invalidate_table_cache_keys(self.parent_cache, self.to_be_invalidated) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/cachalot/utils.py in _invalidate_table_cache_keys
                cache.set_many(d, None) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/redis_cache/backends/single.py in set_many
                        return self._set_many(self.master_client, new_data) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/redis_cache/backends/base.py in _set_many
                    return client.mset(data) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/redis/client.py in mset
                    return self.execute_command('MSET', *items) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/redis/client.py in execute_command
                        return self.parse_response(connection, command_name, **options) ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/redis/client.py in parse_response
                    response = connection.read_response() ...
▶ Local vars
/app/.heroku/python/lib/python2.7/site-packages/redis/connection.py in read_response
                        raise response ...
▶ Local vars
```

**Comment by BertrandBordage** [comment]
That’s because you’re using django-redis-cache and not django-redis.
@nchampsavoir: Is that also your case?

I didn’t see that django-redis-cache was still maintained, it was abandoned for a while, and django-cachalot doesn’t support it. But now that it has more support, I’ll probably support it in django-cachalot, or at least create a check to raise an error if django-redis-cache is used.

However, I’m not sure I’ll support it, because this bug report shows that it’s currently partially broken. The command I use in this case is a normal call of the django cache API, it shouldn’t lead to this error.

**Comment by eressler** [comment]
Ahh interesting. Good to know. I'll switch to using django-redis and see if that resolves the issue.

**Comment by eressler** [comment]
Ahhh! Yep that was the issue, at least for me. Switched to using django-redis instead and the error went away. I only originally chose django-redis-cache since that is Heroku's recommendation.  I might shoot them an email and suggest they change it based on the current status of that project.

Either way, if you choose not to support it a error message might be useful since the current error doesn't really give much info on the cause.  Or perhaps update the docs to list supported django redis modules?

Anyway, thanks for the help in resolving this for me!

**Comment by nchampsavoir** [comment]
Hi.

Yes, I was using django-redis-cache as well. I changed my code and migrated
to memcached a while back, which solved my issue. I'll give django-redis a
try if I get any chance.

Cheers

2015-09-29 21:03 GMT+02:00 Eric Ressler <notifications@github.com>:

**Comment by BertrandBordage** [comment]
Fixed with 64650340b6390f119cd04bb5e98824e760fa0fbe. No django-redis-cache support, but a check verifies that configured databases and caches are valid. If the default database or cache is invalid, an error is raised and the process halts. If a non-default daabase is invalid, a warning is displayed.

**Comment by wanwenjie** [comment]
May be your dict is None
